alphazero/mcts/mcts.py

```diff
@@ -0,0 +1,282 @@
+        Args:
+            game (GameClassType): Game that is currently played
+
+        Returns:
+            result (int): Final result of rollout
+        """
+        state = self._state
+        while not game.is_game_over(state):
+            legal_actions = game.get_legal_moves(state)
+            action = sample_tensor_indices(legal_actions, 1)[0]
+            player = 1 if (state == 0).sum() > (state == 1).sum() else 0
+            state = game.board_after_move(state, player, action)
+        result = game.result(state)
+        return result
+
+    def calc_policy_value(
+        self, rollout: bool, game: GameClassType, network: torch.nn.Module = None
+    ) -> (torch.Tensor, float):
+        """
+        Calculates the policy and value for the node. If rollout performs a rollout, else uses network to get policy and value
+        Args:
+            rollout (bool): Whether to perform a rollout or not
+            game (GameClassType): Game that is played
+            network (torch.nn.Module, optional): Optional, must be set if rollout is false
+
+        Returns:
+            policy (torch.Tensor): The calculated policy
+            value (float): The calculated value
+        """
+        if not rollout:
+            # TODO: Normalize probabilities after filter legal actions
+            self._policy, self._value = network(self._state)
+        else:
+            self._value = self.rollout(game)
+            policy = game.get_legal_moves(self._state)
+            self._policy = policy.float() / policy.sum()
+        return self._policy, self._value
+
+    def expand(self, game: GameClassType):
+        """
+        Expands node
+        Args:
+            game (GameClassType): Game that is played
+        """
+        legal_actions = game.get_legal_moves(self._state)
+        for index in legal_actions.nonzero():
+            action = torch.zeros(legal_actions.size())
+            action[tuple(index)] = 1
+            prior = self._policy[tuple(index)].item()
+            player = 1 if (self._state == 0).sum() > (self._state == 1).sum() else 0
+            child = MCTSNode(
+                state=game.board_after_move(self._state, player, tuple(index)),
+                action=action,
+                prior=prior,
+            )
+            self.add_child(child)
+
+    @property
+    def is_leaf(self):
+        return len(self.children) == 0
+
+    @property
+    def q_value(self):
+        if self.n_visit > 0:
+            return self.total_value / self.n_visit
+        else:
+            return 0
+
+    @property
+    def state(self):
+        return self._state
+
+    @property
+    def action(self):
+        return self._action
+
+    @property
+    def children(self):
+        return self._children
+
+    @property
+    def is_root(self):
+        return self.parent is None
+
+    def u_value(self, sum_n: int, c_puct: float = 1.0):
```

Comment:
can you point me to where this is defined? I'm guessing it's in the alphagozero paper somewhere?